### PR TITLE
🐛 normalize MachineDeployment version validation

### LINKED ISSUE #4651: MachineDeployment version validation
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

In a MachineDeployment resource spec, if you declare a `spec.template.spec.version` value that omits the `v` prefix (e.g., "1.19.10" instead of "v1.19.10") a strange side-effect happens: the MachineDeployment gets stuck in a ScalingUp state:

```
NAME                        PHASE       REPLICAS   READY   UPDATED   UNAVAILABLE
acse-test-capz-40a35-md-0   ScalingUp                                1
```

In the capi-controller logs you see a continual stream of errors like this, suggesting that the MachineDeployment continues to try to recreate itself:

```
E0521 20:36:49.912570       1 machinedeployment_controller.go:145] controller-runtime/manager/controller/machinedeployment "msg"="Failed to reconcile MachineDeployment" "error"="machinesets.cluster.x-k8s.io \"acse-test-capz-40a35-md-0-567db869fd\" already exists" "name"="acse-test-capz-40a35-md-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachineDeployment" 
```

Upon discovering that the version was missing the "v", editing the MachineDeployment to change the version to "v1.19.10" quickly unblocked forward progress:

```
$ k get machinedeployments
NAME                        PHASE     REPLICAS   READY   UPDATED   UNAVAILABLE
acse-test-capz-40a35-md-0   Running   1          1       1
```

Most curious is the fact that the cluster is otherwise running fine, with the node that correlates to the above MachineDeployment in a running state. This edge case was detected upon attempting (and failing) to scale the MachineDeployment out to 3.

**What did you expect to happen:**

I would expect one of the following:

1) Version validation to be enforced consistently across all resource types. In my repro scenario, both the KubeadmControlPlane and MachineDeployment resources had a value of "1.19.10" (missing the "v" prefix). It seems that the "v" prefix is automatically added to the spec for KubeadmControlPlane resources:

```
$ k get kubeadmcontrolplane acse-test-capz-40a35-control-plane -o yaml | grep 'version: v'
  version: v1.19.10
```

I would expect the same to be true for MachineDeployments as well

2) If there is a reason for MachineDeployment version validation to be more strict, I'd expect that validation failure to be clearly expressed in the event and controller logs, for quicker discovery.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: v1alpha4
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`): 1.19.10
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04-LTS

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**:

This PR normalizes the MachineDeployment Kubernetes version validation flows to always include a "v" prefix, yielding a consistent value that matches the pattern `v.<major>.<minor>.<patch>`.

This approach matches what we're doing for Machine resource admission here:

- https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/machine_webhook.go#L125-L129
- https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/machine_webhook.go#L58

In fact, doing this in MachineDeployment fixes a bug that results from Machine and MachineDeployment showing a different version, for example "v1.19.10" and "1.19.10", respectively.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4651

### CODE REVIEW

File: api/v1alpha4/machinedeployment_webhook.go [hunk 1]
```diff
@@ -45,6 +46,11 @@ var _ webhook.Validator = &MachineDeployment{}
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
 func (m *MachineDeployment) Default() {
 	PopulateDefaultsMachineDeployment(m)
+	// tolerate version strings without a "v" prefix: prepend it if it's not there
+	if m.Spec.Template.Spec.Version != nil && !strings.HasPrefix(*m.Spec.Template.Spec.Version, "v") {
+		normalizedVersion := "v" + *m.Spec.Template.Spec.Version
```
Comment (vincepri) [inline api/v1alpha4/machinedeployment_webhook.go:51]: Doesn't this cause a rollout because we're still hashing the Template spec?

Comment (jackfrancis) [reply api/v1alpha4/machinedeployment_webhook.go:51]: Do you mean mutating the value will result in the resource being marked dirty and then a new reconciliation will kick off (in the middle of the existing reconciliation)?

Comment (vincepri) [reply api/v1alpha4/machinedeployment_webhook.go:51]: Yes

Comment (jackfrancis) [reply api/v1alpha4/machinedeployment_webhook.go:51]: Why doesn't the mutation in the properties handled by the `PopulateDefaultsMachineDeployment` func have the same undesirable outcome?

Comment (vincepri) [reply api/v1alpha4/machinedeployment_webhook.go:51]: It technically does, but the implementation of `PopulateDefaultsMachineDeployment` makes sure to only set certain things that are either empty, or forced (which are there from when the object is created.

Comment (CecileRobertMichon) [reply api/v1alpha4/machinedeployment_webhook.go:51]: @vincepri wouldn't this be a one-time thing though, only if the user creates a MD without the v prefix (which as described in #4651 isn't functional today)? Once the template version has the v prefix it won't modify it / cause a rollout

Comment (vincepri) [reply api/v1alpha4/machinedeployment_webhook.go:51]: That makes sense, I was assuming that the version would go through regardless, but it seems that it doesn't.

**Review by vincepri — COMMENTED** [review]
Should we also add a validation piece to this that matches the expectations we have everywhere else. (Have the version always start with `v`)

**Comment by CecileRobertMichon** [comment]
> Should we also add a validation piece to this that matches the expectations we have everywhere else? (Have the version always start with v)

Good call. @jackfrancis here's where that's done for machines: https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/machine_webhook.go#L125-L129

**Comment by jackfrancis** [comment]
@vincepri @CecileRobertMichon added version validation to MachineDeployment as well

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by jackfrancis** [comment]
Just to be clear: validation happens *after* the admission foo that prepends the "v", right? (Otherwise it's arguably superfluous to add that webook admission foo...)

**Comment by vincepri** [comment]
@jackfrancis Yes, that's correct. In the future we should be able to drop the mutation as well, and just roll with validation :)

**Comment by jackfrancis** [comment]
@vincepri cool, thx for confirming. I do think there's benefit to including the "v" for free (adding convenience for users), but possibly at the expense of making it more likely they'll make a mistake later on down the road by hiding the fact that actually, yeah, you need the "v" in the string!

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve

Thanks @jackfrancis!

**Pull request merged**